ragger_tests/application_client/app.py

```diff
@@ -0,0 +1,211 @@
+import hashlib
+from typing import Generator, Callable, Optional
+from contextlib import contextmanager
+
+from ragger.backend.interface import BackendInterface, RAPDU
+from response_unpacker import PK, LedgerAppSettings, LedgerModel, Version, SignedPayloadEd25519, SignedPayloadSecp256k1, Signed, unpack_get_version_response
+from command_sender import CommandSender, InsType
+from request_packer import RequestPacker
+from application_client.curve import C
+
+class App:
+    def __init__(self, backend: BackendInterface) -> None:
+        self.sender = CommandSender(backend)
+    
+    def access_last_async_response(self) -> Optional[RAPDU]:
+        return self.sender.get_optional_async_response()
+
+    def get_version(self) -> Version:
+        rapdu = self.sender.get_version()
+        return unpack_get_version_response(rapdu.data)
+
+    def get_device_id(self) -> str:
+        rapdu = self.sender.get_device_id()
+        return rapdu.data.hex()
+    
+    def get_app_settings(self) -> LedgerAppSettings:
+        rapdu = self.sender.get_app_settings()
+        return LedgerAppSettings.unpack(raw=rapdu.data)
+    
+    def get_device_model(self) -> LedgerModel:
+        rapdu = self.sender.get_device_model()
+        return LedgerModel.unpack(raw=rapdu.data)
+    
+    @contextmanager
+    def __verify_address(
+        self, 
+        curve: C,
+        navigate: Callable[[], None],
+        path: str, 
+    ) -> Generator[RAPDU, None, None]:
+        global maybe
+        with self.sender.send_verify_address(
+            curve=curve,
+            navigate=navigate,
+            path=path,
+        ) as response:
+            maybe = response
+        yield maybe if maybe is not None else self.sender.get_async_response()
+
+    def verify_address(
+        self, 
+        curve: C,
+        path: str, 
+        navigate: Callable[[], None] = lambda: None
+    ) -> str:
+        global response
+        with self.__verify_address(
+            curve=curve,
+            navigate=navigate,
+            path=path
+        ) as res:
+            response = res
+        return response.data.decode('utf-8')
+
+    @contextmanager
+    def __sign_generic(
+        self, 
+        ins: InsType,
+        navigate_path: Callable[[], None],
+        navigate_sign: Callable[[], None],
+        path: str, 
+        payload: bytes
+    ) -> Generator[RAPDU, None, None]:
+        global maybe
+        with self.sender.send_sign_generic(
+            ins,
+            navigate_path=navigate_path,
+            navigate_sign=navigate_sign,
+            path=path,
+            payload=payload
+        ) as response:
+            maybe = response
+        yield maybe if maybe is not None else self.sender.get_async_response()
+
+    def sign_rola(
+        self, 
+        curve: C,
```

Comment:
generic over curve `C`, so we share impl between `Secp256k1` and `Ed25519`.

Parsing is done using `Signed. unpack_signed` which in turn is generic.